src/plugins/intel_gpu/src/plugin/transformations/op/gemm.cpp

```diff
@@ -108,60 +67,42 @@ bool Gemm::visit_attributes(ov::AttributeVisitor &visitor) {
-        auto tshape = target_shape;
-        broadcast.set_broadcast_spec(ov::op::BroadcastType::BIDIRECTIONAL);
-        std::unordered_map<size_t, ov::Tensor> const_data;
-        const_data.emplace(1, ov::Tensor(ov::element::i32, ov::Shape{tshape.size()}, static_cast<void*>(tshape.data())));
-        return ov::op::v3::shape_infer(&broadcast,
-                                       std::vector<ov::PartialShape>{shape, ov::PartialShape(ov::Shape{tshape.size()})},
-                                       ov::make_tensor_accessor(const_data));
-    };
-    auto shape_a_b = (target_shape_a.size() > 1) ? broadcast_shape(shape_a, target_shape_a)[0] : shape_a;
-    auto shape_b_b = (target_shape_b.size() > 1) ? broadcast_shape(shape_b, target_shape_b)[0] : shape_b;
-
-    // reshaped shapes
-    auto reshape_shape = [](const ov::PartialShape shape, const std::vector<int64_t>& output_pattern) {
-        ov::op::v1::Reshape reshape;
-        auto opattern = output_pattern;
-        reshape.set_special_zero(true);
-        std::unordered_map<size_t, ov::Tensor> const_data;
-        const_data.emplace(1, ov::Tensor(ov::element::i64, ov::Shape{opattern.size()}, static_cast<void*>(opattern.data())));
-        return ov::op::v1::shape_infer(&reshape,
-                                       std::vector<ov::PartialShape>{shape, ov::PartialShape(ov::Shape{opattern.size()})},
-                                       ov::make_tensor_accessor(const_data));
-    };
-    auto shape_a_r = (output_pattern_a.size() > 1) ? reshape_shape(shape_a_b, output_pattern_a)[0] : shape_a_b;
-    auto shape_b_r = (output_pattern_b.size() > 1) ? reshape_shape(shape_b_b, output_pattern_b)[0] : shape_b_b;
-
-    // transposed shapes
-    auto transpose_shape = [](const ov::PartialShape shape, const std::vector<int64_t>& order) {
-        auto shape_transposed = ov::PartialShape::dynamic(shape.rank());
+    // transposed shape
+    auto transpose_pshape = [](const ov::PartialShape pshape, const std::vector<int64_t>& order) {
+        auto transposed_pshape = ov::PartialShape::dynamic(pshape.rank());
         for (size_t i = 0; i < order.size(); i++) {
-            shape_transposed[i] = shape[order[i]];
+            transposed_pshape[i] = pshape[order[i]];
         }
 
-        return shape_transposed;
+        return transposed_pshape;
     };
-    auto shape_a_t = (order_a.size() > 1) ? transpose_shape(shape_a_r, order_a) : shape_a_r;
-    auto shape_b_t = (order_b.size() > 1) ? transpose_shape(shape_b_r, order_b) : shape_b_r;
+
+    auto shape_a_t = (order_a.size() > 1) ? transpose_pshape(shape_a, order_a) : shape_a;
+    auto shape_b_t = (order_b.size() > 1) ? transpose_pshape(shape_b, order_b) : shape_b;
+
+    // broadcast all batch dimensions
+    if (shape_a_t.rank().is_static() && shape_a_t.rank().is_static() && (shape_a_t.size() == shape_b_t.size())) {
+        size_t max_rank = shape_a_t.size();
+        for (size_t i = 0; i < max_rank - 2; ++i) {
```

Comment:
IIRC, matmul inputs may have rank = 1. In that case max_rank - 2 will be equal to size_t max

Reply:
Added condition to check rank of the inputs and broadcast batch dims.